File: src/exoticatechnologies/ui/impl/shop/upgrades/methods/RecoverMethod.java
package exoticatechnologies.ui.impl.shop.upgrades.methods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.*;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.impl.campaign.ids.Commodities;
import com.fs.starfarer.api.util.Misc;
import exoticatechnologies.hullmods.ExoticaTechHM;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.modifications.upgrades.Upgrade;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;

import java.util.HashMap;
import java.util.Map;

public class RecoverMethod implements UpgradeMethod {
    @Override
    public String getOptionText(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        int creditCost = getCreditCost(fm, mods, upgrade);

        String creditCostFormatted = Misc.getFormat().format(creditCost);
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOption")
                .format("credits", creditCostFormatted)
                .toString();
    }

    @Override
    public String getOptionTooltip(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOptionTooltip").toString();
    }

    @Override
    public boolean canShow(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        return mods.hasUpgrade(upgrade);
    }

    @Override
    public boolean canUse(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        if (mods.hasUpgrade(upgrade)) {
            int creditCost = getCreditCost(fm, mods, upgrade);
            return Global.getSector().getPlayerFleet().getCargo().getCredits().get() >= creditCost;
        }
        return false;
    }

    @Override
    public String apply(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        CampaignFleetAPI fleet = Global.getSector().getPlayerFleet();

        int creditCost = getCreditCost(fm, mods, upgrade);

        CargoStackAPI stack = Utilities.getUpgradeChip(fleet.getCargo(), upgrade.getKey(), mods.getUpgrade(upgrade));
        if (stack != null) {
            stack.add(1);
        } else {
            fleet.getCargo().addSpecial(upgrade.getNewSpecialItemData(mods.getUpgrade(upgrade)), 1);
        }

        mods.removeUpgrade(upgrade);
        fleet.getCargo().getCredits().subtract(creditCost);

        mods.save(fm);
        ExoticaTechHM.addToFleetMember(fm);

        return StringUtils.getString("UpgradesDialog", "UpgradeRecoveredSuccessfully");
    }

    /**
     * Sums up the floats in the map.
     *
     * @param resourceCosts resource cost map
     * @return The sum.
     */
    private static int getCreditCostForResources(Map<String, Integer> resourceCosts) {
        float creditCost = 0;

        for (Map.Entry<String, Integer> resourceCost : resourceCosts.entrySet()) {
            creditCost += Utilities.getItemPrice(resourceCost.getKey()) * resourceCost.getValue();
        }
        return (int) creditCost;
    }

    private static int getCreditCost(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade) {
        float resourceCreditCost = getCreditCostForResources(upgrade.getResourceCosts(fm, mods.getUpgrade(upgrade)));

        return (int) (resourceCreditCost * 0.166);
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market, boolean hovered) {
        Map<String, Float> resourceCosts = new HashMap<>();

        if (hovered) {
            resourceCosts.put(Commodities.CREDITS, (float) getCreditCost(fm, mods, upgrade));

            String resourceName = StringUtils.getTranslation("ShipListDialog", "ChipName")
                    .format("name", upgrade.getName())
                    .toString();

            if (mods.hasUpgrade(upgrade)) {
                resourceName = StringUtils.getTranslation("ShipListDialog", "UpgradeChipWithLevelText")
                        .format("upgradeName", upgrade.getName())
                        .format("level", mods.getUpgrade(upgrade))
                        .toString();
            }

            resourceCosts.put(String.format("&%s", resourceName), -1f);
        }

        return resourceCosts;
    }

    @Override
    public boolean usesBandwidth() {
        return false;
    }

    @Override
    public boolean usesLevel() {
        return false;
    }
}

File: src/exoticatechnologies/ui/impl/shop/upgrades/methods/RecoverMethod.kt
package exoticatechnologies.ui.impl.shop.upgrades.methods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.ShipVariantAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.impl.campaign.ids.Commodities
import exoticatechnologies.hullmods.ExoticaTechHM
import exoticatechnologies.modifications.ShipModLoader
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.upgrades.Upgrade
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities
import lombok.Getter

class RecoverMethod : UpgradeMethod {
    @Getter
    override var key = "recover"
    override fun getOptionText(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOption")
            .toString()
    }

    override fun getOptionTooltip(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOptionTooltip").toString()
    }

    override fun canShow(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): Boolean {
        return mods.hasUpgrade(upgrade)
    }

    override fun canUse(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): Boolean {
        market ?: return false

        if (mods.hasUpgrade(upgrade)) {
            val creditCost = getCreditCost(member, mods, upgrade)
            return Global.getSector().playerFleet.cargo.credits.get() >= creditCost
        }
        return false
    }

    override fun canUseIfMarketIsNull(): Boolean {
        return false
    }

    override fun apply(member: FleetMemberAPI, variant: ShipVariantAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): String {
        val fleet = Global.getSector().playerFleet
        val creditCost = getCreditCost(member, mods, upgrade)
        val stack = Utilities.getUpgradeChip(fleet.cargo, upgrade.key, mods.getUpgrade(upgrade))
        if (stack != null) {
            stack.add(1f)
        } else {
            fleet.cargo.addSpecial(upgrade.getNewSpecialItemData(mods.getUpgrade(upgrade)), 1f)
        }
        fleet.cargo.credits.subtract(creditCost.toFloat())

        mods.removeUpgrade(upgrade)
        ShipModLoader.set(member, variant, mods)
        ExoticaTechHM.addToFleetMember(member, variant)

        return StringUtils.getString("UpgradesDialog", "UpgradeRecoveredSuccessfully")
    }

    override fun getResourceCostMap(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?,
        hovered: Boolean
    ): Map<String, Float> {
        val resourceCosts: MutableMap<String, Float> = HashMap()
        if (hovered) {
            resourceCosts[Commodities.CREDITS] = getCreditCost(member, mods, upgrade).toFloat()
            var resourceName = StringUtils.getTranslation("ShipListDialog", "ChipName")
                .format("name", upgrade.name)
                .toString()
            if (mods.hasUpgrade(upgrade)) {
                resourceName = StringUtils.getTranslation("ShipListDialog", "UpgradeChipWithLevelText")
                    .format("upgradeName", upgrade.name)
                    .format("level", mods.getUpgrade(upgrade))
                    .toString()
            }
            resourceCosts[String.format("&%s", resourceName)] = -1f
        }
        return resourceCosts
    }

    override fun usesBandwidth(): Boolean {
        return false
    }

    override fun usesLevel(): Boolean {
        return false
    }

    override fun shouldLoad(): Boolean {
        return true
    }

    companion object {
        /**
         * Sums up the floats in the map.
         *
         * @param resourceCosts resource cost map
         * @return The sum.
         */
        private fun getCreditCostForResources(resourceCosts: Map<String, Int>): Int {
            var creditCost = 0f
            for ((key, value) in resourceCosts) {
                creditCost += Utilities.getItemPrice(key) * value
            }
            return creditCost.toInt()
        }

        fun getCreditCost(fm: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade): Int {
            val resourceCreditCost =
                getCreditCostForResources(upgrade.getResourceCosts(fm, mods.getUpgrade(upgrade))).toFloat()
            return (resourceCreditCost * 0.166).toInt()
        }
    }
}